### PR TITLE
Benchmarks for model X parameter set combinations

### PR DESCRIPTION
**Pull request by Vaibhav-Chopra-GT**
# Description

Added benchmarks for `SPM`, `SPMe`, and `DFN` with all the parameter sets (except Sulzer2019 and Xu2019).

Some concerns -
- I could not solve the `DFN` model with `Mohtat2020` parameter sets for the benchmarks (commented code). I tried varying `dt_max` and used different solvers but nothing worked. Here is a reproducible example for the same-
```python
import numpy as np
import pybamm
import matplotlib.pyplot as plt

def prepare_model_Mohtat2020(model):
    geometry = model.default_geometry

    param = pybamm.ParameterValues("Mohtat2020")
    param.process_model(model)
    param.process_geometry(geometry)

    var_pts = {"x_n": 20, "x_s": 20, "x_p": 20, "r_n": 30, "r_p": 30, "y": 10, "z": 10}
    mesh = pybamm.Mesh(geometry, model.default_submesh_types, var_pts)

    disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
    disc.process_model(model)

solver = pybamm.CasadiSolver()
model = pybamm.lithium_ion.DFN()
c_rate = 1
tmax = 4000 / c_rate
nb_points = 500
t_eval = np.linspace(0, tmax, nb_points)
prepare_model_Mohtat2020(model)
solver.solve(model, t_eval=t_eval)
solution = solver.solve(model, t_eval=t_eval)
voltage = solution['Terminal voltage [V]']
c_s_n_surf = solution['Negative particle surface concentration']
c_s_p_surf = solution['Positive particle surface concentration']
t = solution["Time [s]"].entries
x = solution["x [m]"].entries[:, 0]
f, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(13,4))

ax1.plot(t, voltage(t))
ax1.set_xlabel(r'$Time [s]$')
ax1.set_ylabel('Terminal voltage [V]')

ax2.plot(t, c_s_n_surf(t=t, x=x[0]))  
ax2.set_xlabel(r'$Time [s]$')
ax2.set_ylabel('Negative particle surface concentration')

ax3.plot(t, c_s_p_surf(t=t, x=x[-1])) 
ax3.set_xlabel(r'$Time [s]$')
ax3.set_ylabel('Positive particle surface concentration')

plt.tight_layout()
plt.show()
```
The error-
```
The linesearch algorithm failed with too small a step.
At t = 3.7305e-005 and h = 3.3406e-019, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.73046e-005 and h = 3.12482e-019, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.73044e-005 and h = 3.16429e-020, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.7305e-005 and h = 3.3406e-019, the corrector convergence failed repeatedly or with |h| = hmin.
Traceback (most recent call last):
  File "d:\LearningProjects\python\second.py", line 25, in <module>
    solver.solve(model, t_eval=t_eval)
  File "C:\Users\Vaibhav Chopra\AppData\Local\Programs\Python\Python39\lib\site-packages\pybamm\solvers\base_solver.py", line 1018, in solve
    new_solution = self._integrate(
  File "C:\Users\Vaibhav Chopra\AppData\Local\Programs\Python\Python39\lib\site-packages\pybamm\solvers\casadi_solver.py", line 263, in _integrate
    raise pybamm.SolverError(
pybamm.expression_tree.exceptions.SolverError: Maximum number of decreased steps occurred at t=0.0. Try solving the model up to this time only or reducing dt_max (currently, dt_max=5.010622778599846).
PS D:\LearningProjects> python -u "d:\LearningProjects\python\second.py"
The linesearch algorithm failed with too small a step.
At t = 3.7305e-005 and h = 3.3406e-019, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.73046e-005 and h = 3.12482e-019, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.73044e-005 and h = 3.16429e-020, the corrector convergence failed repeatedly or with |h| = hmin.
At t = 3.7305e-005 and h = 3.3406e-019, the corrector convergence failed repeatedly or with |h| = hmin.
Traceback (most recent call last):
  File "d:\LearningProjects\python\second.py", line 25, in <module>
    solver.solve(model, t_eval=t_eval)
  File "C:\Users\Vaibhav Chopra\AppData\Local\Programs\Python\Python39\lib\site-packages\pybamm\solvers\base_solver.py", line 1018, in solve
    new_solution = self._integrate(
  File "C:\Users\Vaibhav Chopra\AppData\Local\Programs\Python\Python39\lib\site-packages\pybamm\solvers\casadi_solver.py", line 263, in _integrate
    raise pybamm.SolverError(
pybamm.expression_tree.exceptions.SolverError: Maximum number of decreased steps occurred at t=0.0. Try solving the model up to this time only or reducing dt_max (currently, dt_max=5.010622778599846).
```
- Should I group these benchmarks in some way. For example- multiple files with each file containing benchmarks of a particular parameter or model.
- As discussed with @brosaplanella on slack I have used `ScikitsDaeSolver` for some benchmarks, will this create a problem or is it safe to use different solvers for similar benchmarks?

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: benchmarks/time_solve_models.py [hunk 1]
```diff
@@ -22,11 +22,328 @@ def prepare_model(model):
     disc.process_model(model)
 
 
+def prepare_model_ORegan2021(model):
+    geometry = model.default_geometry
+
+    # load parameter values and process model and geometry
+    param = pybamm.ParameterValues("ORegan2021")
```
Comment (brosaplanella) [inline benchmarks/time_solve_models.py:29]: If you use parameters (as in my previous comment) you can have a single `prepare_model(model, parameters)` function.

File: benchmarks/time_solve_models.py [hunk 2]
```diff
@@ -22,11 +22,328 @@ def prepare_model(model):
...
+    disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
+    disc.process_model(model)
+
+
 def solve_model_once(model, solver, t_eval):
     solver.solve(model, t_eval=t_eval)
 
 
-class TimeSolveSPM:
+class TimeSolveSPMMarquis2019:
```
Comment (brosaplanella) [inline benchmarks/time_solve_models.py:182]: Same as before, but need to be careful as you will have two `parameter` sets for the tests: the battery parameters and the solve (True, False). In the [docs](https://asv.readthedocs.io/en/stable/writing_benchmarks.html#parameterized-benchmarks) there is an example on how to do it.

File: benchmarks/time_solve_models.py [hunk 3]
```diff
@@ -22,11 +22,328 @@ def prepare_model(model):
...
+        c_rate = 1
+        tmax = 4000 / c_rate
+        nb_points = 500
+        self.t_eval = np.linspace(0, tmax, nb_points)
+        prepare_model_Ramadass2004(self.model)
+        if solve_first:
+            solve_model_once(self.model, TimeSolveSPMRamadass2004.solver, self.t_eval)
+
+    def time_solve_model(self, solve_first):
+        TimeSolveSPMRamadass2004.solver.solve(self.model, t_eval=self.t_eval)
```
Comment (brosaplanella) [inline benchmarks/time_solve_models.py:287]: This applies everywhere (but just spotted it here). You can instead type `self` to refer to the own class, so it is more compact.

File: benchmarks/time_solve_models.py [hunk 4]
```diff
@@ -36,15 +353,181 @@ def setup(self, solve_first):
...
+        if solve_first:
+            solve_model_once(self.model, TimeSolveSPMePrada2013.solver, self.t_eval)
+
+    def time_solve_model(self, solve_first):
+        TimeSolveSPMePrada2013.solver.solve(self.model, t_eval=self.t_eval)
+
+
+class TimeSolveSPMeAi2020:
+    params = [True, False]
+    solver = pybamm.ScikitsDaeSolver()
```
Comment (brosaplanella) [inline benchmarks/time_solve_models.py:438]: Sorry, I didn't explain myself very well on Slack. In the benchmarks we should use the same solver (with same settings, like dt_max) for all tests, so results are comparable. I suggested trying Scikits to check if the model solved fine. If it does, then you can go and fiddle with CasADI. We still don't know why this happens, but sometimes changing the tmax (like to `3800 / c_rate` or `4200 / c_rate`) makes the solver work. If you can't get it to work straight away, just comment the benchmark out and we can skip it for the time being.

**Review by brosaplanella — CHANGES_REQUESTED** [review]
Hi Vaibhav! It looks pretty good, but there are some changes I suggested (mostly to make use of parameters in benchmarks so you can simplify your code a lot). 

I couldn't attend last meeting, so if any of my comments contradicts anything you agreed previously do let me know (and apologies in advanced).

File: pybamm/models/submodels/interface/sei/sei_growth.py [hunk 5]
```diff
@@ -84,9 +84,15 @@ def get_coupled_variables(self, variables):
         if self.options["SEI"] == "reaction limited":
             # alpha = param.alpha
             C_sei = param.C_sei_reaction
+            T = variables["Negative electrode temperature"]
```
[the remark this answers is not in the document]
Comment (Vaibhav-Chopra-GT) [reply pybamm/models/submodels/interface/sei/sei_growth.py]: Yes, done!

**Comment by Vaibhav-Chopra-GT** [comment]
Thank you for the review @tinosulzer, @brosaplanella! I have parameterized the benchmarks and I am still looking into the solvers. Could you please review it again and give your feedback. Also, keeping the voltage same for every parameter did not throw any error, should I keep it the same or change it to match every parameter's lower voltage.

**Review by brosaplanella — APPROVED** [review]
Nice, it looks much cleaner now!

**Comment by brosaplanella** [comment]
> Thank you for the review @tinosulzer, @brosaplanella! I have parameterized the benchmarks and I am still looking into the solvers. Could you please review it again and give your feedback. Also, keeping the voltage same for every parameter did not throw any error, should I keep it the same or change it to match every parameter's lower voltage.

For the voltage, it would be better to test down to the right voltage specified in the parameters so we ensure we have a full discharge. You can just get the value for `Lower voltage cut-off [V]' and format the string of the experiment.

File: benchmarks/time_setup_models_and_sims.py [hunk 6]
```diff
@@ -89,12 +149,24 @@ def time_setup_SPMe_simulation(self, with_experiment):
...
+            "ORegan2021",
+            "NCA_Kim2011",
+            "Prada2013",
+            "Ai2020",
+            "Ramadass2004",
+            "Mohtat2020",
+            "Chen2020",
+            "Chen2020_plating",
+            "Ecker2015",
+        ],
```
Comment (priyanshuone6) [inline benchmarks/time_setup_models_and_sims.py:165]: Could you make a global variable with a list of these parameter sets outside the classes and use that in all the functions to reduce code duplication

**Review by tinosulzer — APPROVED** [review]
Thanks, looks good!

**Comment by priyanshuone6** [comment]
Let’s merge this

**Pull request merged**